apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicy.cs

```diff
@@ -0,0 +1,315 @@
+            private static readonly IPostPolicyElement[] s_knownElements = new IPostPolicyElement[]
+            {
+                Acl.Element,
+                BucketName.Element,
+                CacheControl.Element,
+                ContentDisposition.Element,
+                ContentEncoding.Element,
+                ContentLength.Element,
+                ContentType.Element,
+                ObjectExpiry.Element,
+                Key.Element,
+                SuccessActionRedirect.Element,
+                SuccessActionStatus.Element
+            };
+
+            private static readonly ICustomElementFactory[] s_customElementFactories = new ICustomElementFactory[]
+            {
+                GoogleMetadataElementFactory.Default,
+                CustomElementFactory.Default
+            };
+
+            private readonly IDictionary<IPostPolicyElement, IPostPolicyCondition> _conditions =
+                new Dictionary<IPostPolicyElement, IPostPolicyCondition>();
+
+            private PostPolicy() { }
+
+            internal PostPolicy(PostPolicy other)
+            {
+                GaxPreconditions.CheckNotNull(other, nameof(other));
+                foreach (var condition in other._conditions)
+                {
+                    _conditions.Add(condition);
+                }
+            }
+
+            /// <summary>
+            /// Creates a new <see cref="PostPolicy"/> instance which requieres
+            /// an exact match for the <see cref="Key"/> element.
+            /// </summary>
+            /// <param name="exactKeyValue">The value to match against. Must not be null.</param>
+            /// <returns>A new <see cref="PostPolicy"/> instance.</returns>
+            public static PostPolicy FromExactKey(string exactKeyValue)
+            {
+                var postPolicy = new PostPolicy();
+                postPolicy.SetField(PostPolicyKnownElement.Key, exactKeyValue);
+                return postPolicy;
+            }
+
+            /// <summary>
+            /// Creates a new <see cref="PostPolicy"/> instance which requieres
+            /// a start with match for the <see cref="Key"/> element.
+            /// </summary>
+            /// <param name="keyPrefix">The value to match against. Must not be null.</param>
+            /// <returns>A new <see cref="PostPolicy"/> instance.</returns>
+            public static PostPolicy FromKeyStartsWith(string keyPrefix)
+            {
+                var postPolicy = new PostPolicy();
+                postPolicy.SetStartsWith(PostPolicyKnownElement.Key, keyPrefix);
+                return postPolicy;
+            }
+
+            internal string Bucket =>
+                _conditions.TryGetValue(BucketName.Element, out IPostPolicyCondition condition) &&
+                condition is ExactMatch<BucketName, string> exactMatch
+                ? exactMatch.Value : null;
+
+            /// <summary>
+            /// Collection of fields as defined by this policy.
```

Comment:
It feels like the name and first summary should emphasize the "exact match" part. Do we need this method at all? (I see the StartsWithConditions etc have been removed - could this go too?)